File: python_version/cli/commands/sync.py
import os
import subprocess
from datetime import datetime

from cli.utils.ui import print_command_banner, print_success, print_error, print_info, console

def run_command(command, cwd=None):
    """Run a shell command and return its success status and output"""
    try:
        result = subprocess.run(
            command, 
            cwd=cwd, 
            shell=True, 
            check=False,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True
        )
        return result.returncode == 0, result.stdout, result.stderr
    except Exception as e:
        return False, "", str(e)

def main(config: dict):
    """Sync LeetCode workspace with remote git repository"""
    print_command_banner("Sync LeetCode Workspace")
    
    base_dir = config.get("base_dir")
    
    if not base_dir or not os.path.exists(base_dir):
        print_error(f"Base directory '{base_dir}' does not exist.")
        return
        
    # Check if base_dir is a git repository
    is_git, _, _ = run_command("git rev-parse --is-inside-work-tree", cwd=base_dir)
    if not is_git:
        print_error(f"'{base_dir}' is not a Git repository.")
        print_info("Please initialize git and set up a remote repository first:")
        print_info(f"  cd {base_dir}")
        print_info("  git init")
        print_info("  git remote add origin <your-repo-url>")
        return
        
    print_info(f"Syncing workspace: {base_dir}")
    console.print()
    
    # 1. Add all changes
    print_info("Staging changes...")
    success, out, err = run_command("git add .", cwd=base_dir)
    if not success:
        print_error("Failed to stage changes.")
        if err: console.print(err)
        return
        
    # Check if there are changes to commit
    has_changes, out, _ = run_command("git status --porcelain", cwd=base_dir)
    if not out.strip():
        print_info("No changes to sync. Workspace is up to date.")
        return
        
    # 2. Commit changes
    timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    commit_msg = f"LeetCode Sync: {timestamp}"
    print_info(f"Committing changes...")
    
    success, out, err = run_command(f'git commit -m "{commit_msg}"', cwd=base_dir)
    if not success:
        print_error("Failed to commit changes.")
        if err: console.print(err)
        return
        
    # 3. Pull latest changes (optional, but good practice)
    print_info("Pulling latest changes from remote (if any)...")
    run_command("git pull --rebase", cwd=base_dir) # Don't fail if pull fails (might not have tracking branch)
        
    # 4. Push changes
    print_info("Pushing to remote...")
    success, out, err = run_command("git push", cwd=base_dir)
    if not success:
        print_error("Failed to push changes.")
        print_info("You might need to set upstream branch manually the first time:")
        print_info("  cd " + base_dir)
        print_info("  git push -u origin main")
        if err: console.print(err)
        return
        
    console.print()
    print_success("Successfully synced LeetCode workspace!")

File: go-cli/internal/commands/sync.go
package commands

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"

	"leetcli/internal/config"
	"leetcli/internal/template"
)

func Sync(args []string, cfg *config.Config, ui UI) {
	ui.WriteOutput(MsgPlain, "--- Sync LeetCode Workspace ---\n")

	baseDir := cfg.BaseDir
	if baseDir == "" {
		ui.WriteOutput(MsgError, "Base directory not configured.")
		return
	}
	if _, err := os.Stat(baseDir); os.IsNotExist(err) {
		ui.WriteOutput(MsgError, "Base directory '%s' does not exist.", baseDir)
		return
	}

	gitDir := filepath.Join(baseDir, ".git")
	if _, err := os.Stat(gitDir); os.IsNotExist(err) {
		ui.WriteOutput(MsgError, "'%s' is not a Git repository.", baseDir)
		ui.WriteOutput(MsgInfo, "Please initialize git and set up a remote repository first:")
		ui.WriteOutput(MsgPlain, "  cd %s", baseDir)
		ui.WriteOutput(MsgPlain, "  git init")
		ui.WriteOutput(MsgPlain, "  git remote add origin <your-repo-url>")
		return
	}

	ui.WriteOutput(MsgInfo, "Syncing workspace: %s", baseDir)

	ui.WriteOutput(MsgInfo, "Updating root README.md index...")
	if _, count, err := template.GenerateRootReadme(baseDir, cfg); err == nil {
		ui.WriteOutput(MsgSuccess, "Updated README.md index with %d problem(s).", count)
	} else {
		ui.WriteOutput(MsgError, "Warning: Failed to update root README.md: %v", err)
	}

	ui.WriteOutput(MsgInfo, "Staging changes...")
	if out, err := runGit(baseDir, "add", "."); err != nil {
		ui.WriteOutput(MsgError, "Failed to stage changes: %v", err)
		if out != "" {
			ui.WriteOutput(MsgPlain, "%s", out)
		}
		return
	}

	statusOut, _ := runGit(baseDir, "status", "--porcelain")
	if strings.TrimSpace(statusOut) == "" {
		ui.WriteOutput(MsgInfo, "No changes to sync. Workspace is up to date.")
		return
	}

	timestamp := time.Now().Format("2006-01-02 15:04:05")
	commitMsg := fmt.Sprintf("LeetCode Sync: %s", timestamp)
	ui.WriteOutput(MsgInfo, "Committing changes...")
	if out, err := runGit(baseDir, "commit", "-m", commitMsg); err != nil {
		ui.WriteOutput(MsgError, "Failed to commit changes: %v", err)
		if out != "" {
			ui.WriteOutput(MsgPlain, "%s", out)
		}
		return
	}

	ui.WriteOutput(MsgInfo, "Pulling latest changes from remote (if any)...")
	runGit(baseDir, "pull", "--rebase")

	ui.WriteOutput(MsgInfo, "Pushing to remote...")
	if out, err := runGit(baseDir, "push"); err != nil {
		ui.WriteOutput(MsgError, "Failed to push changes.")
		ui.WriteOutput(MsgInfo, "You might need to set upstream branch manually:")
		ui.WriteOutput(MsgPlain, "  cd %s", baseDir)
		ui.WriteOutput(MsgPlain, "  git push -u origin main")
		if out != "" {
			ui.WriteOutput(MsgPlain, "%s", out)
		}
		return
	}

	ui.WriteOutput(MsgSuccess, "Successfully synced LeetCode workspace!")
}

func runGit(dir string, args ...string) (string, error) {
	cmd := exec.Command("git", args...)
	cmd.Dir = dir
	out, err := cmd.CombinedOutput()
	return string(out), err
}


